core/interval/int.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package interval
+
+import (
+	"sort"
+)
+
+type IntegerInterval struct {
+	Lower int
+	Upper int
+}
+
+func NewIntegerInterval(a, b int) IntegerInterval {
+	return IntegerInterval{
+		Lower: min(a, b),
+		Upper: max(a, b),
+	}
+}
+
+func (i IntegerInterval) Intersects(other IntegerInterval) bool {
+	return i.Lower <= other.Upper && i.Upper >= other.Lower
+}
+
+func (i IntegerInterval) Adjacent(other IntegerInterval) bool {
+	return i.Upper+1 == other.Lower || i.Lower-1 == other.Upper
+}
+
+func (i IntegerInterval) IsSubsetOf(other IntegerInterval) bool {
+	return i.Lower >= other.Lower && i.Upper <= other.Upper
+}
+
+func (i IntegerInterval) Difference(other IntegerInterval) IntegerIntervals {
```

Comment:
Would be good to have a precise definition of what difference is here.

Reply:
included in docstring